presto-hive/src/main/java/io/prestosql/plugin/hive/s3/PrestoS3FileSystem.java

```diff
@@ -775,7 +778,8 @@ private static AWSCredentialsProvider getCustomAWSCredentialsProvider(URI uri, C
         private final Duration maxBackoffTime;
         private final Duration maxRetryTime;
 
-        private boolean closed;
+        private final AtomicBoolean closed = new AtomicBoolean();
```

Comment:
`volatile` ?

anyway, how this relates to `this eliminates reading extra data from S3 by always requesting the
exact range needed`? i didn't dig into the code here, but how range reads impact concurrency? or is it two separate logical changes?

Reply:
We tend to use atomics in preference to volatile. I personally think it's cleaner, since the concurrency aspects are explicit in the API, but I don't feel that strongly about it.

The positioned read methods in `PositionedReadable` are thread-safe. You can open the file once and then call these concurrently from different threads. The default implementation in `FSInputStream` does this via synchronization and seeking (which is a bit broken since it's not safe to use concurrently with normal streaming reads).

I suppose that we could simply ignore the closed state in the read method since it doesn't rely on any mutable state in the object, but it seems better to preserve the expected behavior.